Userland/Utilities/json.cpp

```diff
@@ -121,33 +149,181 @@ void print(JsonValue const& value, int spaces_per_indent, int indent, bool use_c
+                matches.append(array.at(i));
+        }
+        return matches;
+    }
+
+private:
+    int m_start = 0;
+    int m_end = 0;
+    int m_step = 1;
+};
+
+ErrorOr<JsonValue> query(JsonValue const& value, Vector<StringView>& key_parts, size_t key_index)
+{
+    if (key_index == key_parts.size())
+        return value;
+    auto key_str = key_parts[key_index++];
+
+    if (is_single_value_key(key_str)) {
+        auto key = SingleValueKey::parse(key_str);
+        auto result = TRY(key->select(value));
+        return query(result, key_parts, key_index);
+    } else {
+        auto key = TRY(MultiValueKey::parse(key_str));
+        Vector<JsonValue> results;
+        for (auto const& match : key->select(value)) {
+            auto result = TRY(query(match, key_parts, key_index));
+            results.append(result);
         }
-        return JsonValue(JsonArray(matches));
+        return JsonValue(JsonArray(results));
     }
+}
+
+// Normalize negative indices, e.g. an index of -1 would wrap to the last element.
+int normalize_index(int i, size_t length) { return (i < 0) ? length + i : i; }
+
+bool is_single_value_key(StringView key)
+{
+    if (key == "*"sv || key == "[*]"sv)
+        return false;
+    if (key.starts_with('[') && key.ends_with(']')) {
+        auto contains_quote = key.contains('"') || key.contains('\'');
+        auto is_filter = key[1] == '?';
+        return contains_quote && !is_filter;
+    }
+    return true;
+}
+
+NonnullRefPtr<SingleValueKey> SingleValueKey::parse(StringView key)
+{
+    if (key.starts_with('[') && key.ends_with(']'))
+        key = key.substring_view(1, key.length() - 2);
+
+    if ((key.starts_with('\'') && key.ends_with('\''))
+        || (key.starts_with('"') && key.ends_with('"')))
+        key = key.substring_view(1, key.length() - 2);
+
+    return make_ref_counted<SingleValueKey>(key);
+}
 
+ErrorOr<JsonValue> SingleValueKey::select(JsonValue const& value)
+{
     JsonValue result {};
     if (value.is_object()) {
-        result = value.as_object().get(key);
+        result = value.as_object().get(m_key);
     } else if (value.is_array()) {
-        auto key_as_index = key.to_int();
-        if (key_as_index.has_value())
-            result = value.as_array().at(key_as_index.value());
+        auto const& array = value.as_array();
+        auto key_as_index = m_key.to_int();
+        if (!key_as_index.has_value())
+            return Error::from_string_view("Invalid array index"sv);
+        result = array.at(normalize_index(key_as_index.value(), array.size()));
+    }
+    return result;
+}
+
+ErrorOr<NonnullRefPtr<MultiValueKey>> MultiValueKey::parse(StringView const& key)
+{
+    if (key == "*"sv || key == "[*]"sv)
+        return WildcardKey::parse(key);
+    if (key.starts_with('[') && key.ends_with(']'))
+        return TRY(SliceKey::parse(key));
+    VERIFY_NOT_REACHED();
+}
+
+Vector<StringView> split_dotted_key(StringView dotted_key)
```

Comment:
This is not consistent, see 
`MultiValueKey::parse(StringView const& key)` and `split_dotted_key(StringView dotted_key)`.
Both deal with immutable input, so this should be both either `StringView const& ` or `StringView` (if it is implemented similar to an immutable view like `std::string_view`). If so, pass `StringView` by value.